Feature/tn rickandmorty listing page service

tolganacar opened the pull request.

tolganacar committed:
created service and model

tolganacar committed:
Add rick and morty listeleme sayfası mvvm classları eklendi
Add splash screen eklendi

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/adapter/RickMortyAdapter.kt:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    override fun onBindViewHolder(holder: RickMortyHolder, position: Int) {
+        holder.binding.recyclerName.text = characterList.get(position).name
```
holder.binding.apply diyip içerisini düzenleyebilirsin. 

Sürekli olarak "holder.binding." yazarsan okunabilirlik azalır. Bu boilerplate kodu(kalıp kod) .apply {} bloğu ile kaldırabiliriz. Fazlalık koddan kurtulunca daha temiz bir görüntü oluşur.

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/adapter/RickMortyAdapter.kt:
```diff
@@ -0,0 +1,44 @@
+        holder.binding.recyclerName.text = characterList.get(position).name
+        holder.binding.recyclerStatus.text = characterList.get(position).status
+        holder.binding.cardView.setOnClickListener {
+            val action = FeedFragmentDirections.actionFeedFragmentToDetailsFragment()
```
`holder.binding.cardView.setOnClickListener` buradaki gibi listenerların aksiyonlarını adapter içerisine koymayız. Adapter yalnızca modeli xml'e bind eder ve kullanıcının aldığı aksiyonları view'a bildirir. Burada onclick listener setleyebilirsin ama click aksiyonunu adapter dinlemez.

Listenerlerın bulunması gereken yer view yani bu projede Fragment tarafı. Bunu da bir interface veya high order function ile yapabilirsin.

https://github.com/kkocaburak/crypto-currency-mvvm-android/blob/master/crypto-currency-app/app/src/main/java/com/bkarakoca/cryptocurrencyapp/scene/crypto/cryptolist/CryptoCoinListAdapter.kt

https://github.com/kkocaburak/crypto-currency-mvvm-android/blob/master/crypto-currency-app/app/src/main/java/com/bkarakoca/cryptocurrencyapp/scene/crypto/cryptolist/CryptoCoinListFragment.kt

Örnek için yukarıdaki linklere bakabilirsin.

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/adapter/RickMortyAdapter.kt:
```diff
@@ -0,0 +1,44 @@
+import com.tolganacar.rickmorty.util.placeholderProgressBar
+import com.tolganacar.rickmorty.view.FeedFragmentDirections
+
+class RickMortyAdapter(val characterList: ArrayList<Character>): RecyclerView.Adapter<RickMortyAdapter.RickMortyHolder>() {
```
class adını `RMCharacterAdapter` olarak güncelleyebilirsin. `RickMortyAdapter` diyince bu adapter RickMorty e ait herhangi bir şeyin adapteri olabilir.

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/adapter/RickMortyAdapter.kt:
```diff
@@ -0,0 +1,44 @@
+
+class RickMortyAdapter(val characterList: ArrayList<Character>): RecyclerView.Adapter<RickMortyAdapter.RickMortyHolder>() {
+
+    class RickMortyHolder(val binding : RecyclerRowBinding): RecyclerView.ViewHolder(binding.root){
```
aynı şekilde `RMCharacterViewHolder` diyebilirsin

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/model/Character.kt:
```diff
@@ -0,0 +1,30 @@
+    val results: List<Character>
+)
+
+data class Character(
```
`RMCharacter` olarak güncelleyebilirsin.

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/model/Character.kt:
```diff
@@ -0,0 +1,30 @@
+    val created: String
+)
+
+data class Origin(
```
`RMCharacterOrigin`

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/model/Character.kt:
```diff
@@ -0,0 +1,30 @@
+    val url: String
+)
+
+data class Location(
```
`RMCharacterLocation`

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/service/RickMortyAPI.kt:
```diff
@@ -0,0 +1,12 @@
+interface RickMortyAPI {
+
+    @GET("character")
+    fun getCharacters() : Single<RMCharacterResponseModel>
```
metod adını `getRMCharacters()` olarak güncelleyebilirsin

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/service/RickMortyAPIService.kt:
```diff
@@ -0,0 +1,23 @@
+                    .build()
+                    .create(RickMortyAPI::class.java)
+
+    fun getCharacterList(): Single<RMCharacterResponseModel>{
```
`getRMCharacters()`

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/view/DetailsFragment.kt:
```diff
@@ -1,13 +1,18 @@
+import com.tolganacar.rickmorty.R
+import com.tolganacar.rickmorty.viewmodel.DetailsViewModel
 
 class DetailsFragment : Fragment() {
```
`RMCharacterDetailFragment`

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/view/FeedFragment.kt:
```diff
@@ -0,0 +1,85 @@
+import com.tolganacar.rickmorty.viewmodel.FeedViewModel
+import kotlinx.android.synthetic.main.fragment_feed.*
+
+class FeedFragment : Fragment() {
```
`RMCharacterListFragment`

kkocaburak replied:
`RMCharacterListFragment`

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/view/FeedFragment.kt:
```diff
@@ -0,0 +1,85 @@
+        return inflater.inflate(R.layout.fragment_feed, container, false)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
```
buradaki işlemlerini metodlara bölebilirsin:

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/view/FeedFragment.kt:
```diff
@@ -0,0 +1,85 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        viewModel = ViewModelProviders.of(this).get(FeedViewModel::class.java)
```
viewmodel işlemlerini -> `private fun initializeViewModel()` metoduna taşıyabilirsin

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/view/FeedFragment.kt:
```diff
@@ -0,0 +1,85 @@
+        viewModel = ViewModelProviders.of(this).get(FeedViewModel::class.java)
+        viewModel.refreshData()
+
+        recyclerView.layoutManager = LinearLayoutManager(context)
```
aynı şekilde `private fun initializeRecyclerview()`

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/view/FeedFragment.kt:
```diff
@@ -0,0 +1,85 @@
+        recyclerView.layoutManager = LinearLayoutManager(context)
+        recyclerView.adapter = rickMortyAdapter
+
+        observeLiveData()
```
mesela `observeLiveData()` güzel bir iş bölümü olmuş

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/view/FeedFragment.kt:
```diff
@@ -0,0 +1,85 @@
+
+        observeLiveData()
+
+        swipeRefreshLayout.setOnRefreshListener {
```
`private fun setSwipeRefreshLayout()`

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/viewmodel/FeedViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+import io.reactivex.observers.DisposableSingleObserver
+import io.reactivex.schedulers.Schedulers
+
+class FeedViewModel: ViewModel() {
```
`RMCharacterListVM`

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/viewmodel/DetailsViewModel.kt:
```diff
@@ -0,0 +1,14 @@
+import androidx.lifecycle.ViewModel
+import com.tolganacar.rickmorty.model.Character
+
+class DetailsViewModel: ViewModel() {
```
`RMCharacterDetailViewModel`

kkocaburak replied:
class isimlerin uzun olabilir, önemli olan class adını okuduğunda ne yaptıklarını çok net anlayabilmen

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/viewmodel/FeedViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+
+class FeedViewModel: ViewModel() {
+    val characterList = MutableLiveData<List<Character>>()
+    val errorMessage = MutableLiveData<Boolean>()
```
`errorMessage` yerine `shouldShowErrorMessage` diyebilirsin. Boolean değişkenlerde genellikle isEmpty, shouldShow gibi soru belirten değişken isimleri kullanılıyor. 

is-XXX / should-XXX veya aklına gelen diğer soru kalıpları.

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/viewmodel/FeedViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+class FeedViewModel: ViewModel() {
+    val characterList = MutableLiveData<List<Character>>()
+    val errorMessage = MutableLiveData<Boolean>()
+    val loadingBar = MutableLiveData<Boolean>()
```
`isLoading`

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/viewmodel/FeedViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+    private val rickMortyApiService = RickMortyAPIService()
+    private val disposable = CompositeDisposable()
+
+    fun refreshData(){
```
refreshData() ama hangi data :D.

`fetchRMCharacterList()` diyebilirsin

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/viewmodel/FeedViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+    private val disposable = CompositeDisposable()
+
+    fun refreshData(){
+        getDataFromAPI()
```
`getData` ama hangi data :D.

Burada metod içinde de bir mantık hatası olmuş. Metodun amacı datayı yenilemek, içerisinde de buna uygun kod barındırması gerekiyor. Şu anda bu metodun tek yaptığı işlem başka bir metodu çağırmak(`getDataFromAPI()`).

`getDataFromAPI` içindeki kodları bu metod içine alabilirsin.

kkocaburak commented on RickMorty/app/src/main/java/com/tolganacar/rickmorty/viewmodel/FeedViewModel.kt:
```diff
@@ -0,0 +1,48 @@
+            )
+        )
+    }
+}
```
Kullanmadığın png ve jpg dosyalarını silebilirsin. Eğer bir image yükleyeceksende ya svg ya da png formatı olması gerekiyor. jpg lerin boyutu genellikle yüksek olduğu için onları projelere eklemiyoruz.

tolganacar committed:
Edited the holder.binding within the adapter class
Updated some class and function names
Defined the codes under RMCharacterListFragment (onviewcreated) inside new functions
Deleted unused image files

tolganacar committed:
Created onClickListener
Added databinding

tolganacar committed:
Fix pr comments
